### PR TITLE
[webkitcorepy] Support multiple autoinstallers running in parallel

### PR DESCRIPTION
**Pull request by JonWBedard**
#### caf109704d373d1a9ad7ec3bc6caa23c6ab7a0b1
<pre>
[webkitcorepy] Support multiple autoinstallers running in parallel
<a href="https://bugs.webkit.org/show_bug.cgi?id=256933">https://bugs.webkit.org/show_bug.cgi?id=256933</a>
rdar://109488201

Reviewed by Alexey Proskuryakov.

Lay down a lockfile so that only one autoinstaller process can modify an autoinstall
location at a time.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py:
(Package.install): While installing a package, acquire the lock to prevent
other processes from installing packages while this one is.
(AutoInstall): Define manifest.json and lockfile as class members.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
(FileLock.acquire): Disable patch in Timeout to avoid infinite recursion.

Canonical link: <a href="https://commits.webkit.org/264322@main">https://commits.webkit.org/264322@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4aeb5c647cf0ad4306cc5bb44225faaec3667380

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/7414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7675 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7848 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7619 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9588 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7598 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10504 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8642 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/6867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9155 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/7528 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/5952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/6746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/6860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10147 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7350 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/7387 "Failed to checkout and rebase branch from PR 13994") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6699 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10907 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/859 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7085 "Built successfully") | | |

### CODE REVIEW

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py [hunk 1]
```diff
@@ -285,116 +288,129 @@ def install(self):
...
-                    # If a package has a setup.cfg, we can just lay down a dummy setup.py and let setuptools handle the rest
-                    if os.path.exists(os.path.join(candidate, 'setup.cfg')):
-                        with open(os.path.join(candidate, 'setup.py'), 'w') as setup_py:
-                            setup_py.write('from setuptools import setup\n')
-                            setup_py.write('if __name__ == "__main__":\n')
-                            setup_py.write('    setup()\n')
-                    else:
-                        continue
+        lockfile = os.path.join(AutoInstall.directory, AutoInstall.LOCK_FILE)
+        if os.path.exists(lockfile) and time.time() - os.path.getmtime(lockfile) > AutoInstall.LOCKFILE_TIMEOUT:
```
Comment (aproskuryakov) [inline Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: This wouldn't be necessary if `FileLock` used `O_EXLOCK`. Those are cleaned up by the kernel when processes terminate for any reason.

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py [hunk 2]
```diff
@@ -285,116 +288,129 @@ def install(self):
...
+            os.unlink(lockfile)
 
-                AutoInstall.log('Installing {}...'.format(archive))
+        with FileLock(lockfile, timeout=AutoInstall.LOCKFILE_TIMEOUT):
+            # Re-load our manifest, in case another process is using this autoinstall location
```
Comment (aproskuryakov) [inline Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: WebKit style is to have periods at the end of sentences in comments. 

It would also help to spell out the reason more clearly - why exactly is it necessary to reload? Is this because the other process could have already installed some of the packages that we needed to install?

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py [hunk 3]
```diff
@@ -499,8 +518,9 @@ def set_directory(cls, directory):
             cls.userspace_should_own(creation_root)
 
         try:
-            with open(os.path.join(directory, 'manifest.json'), 'r') as file:
-                cls.manifest = json.load(file)
+            with FileLock(os.path.join(directory, AutoInstall.LOCK_FILE), timeout=AutoInstall.LOCKFILE_TIMEOUT):
+                with open(os.path.join(directory, cls.MANIFEST_JSON), 'r') as file:
+                    cls.manifest = json.load(file)
```
Comment (aproskuryakov) [inline Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: Why is this needed? We are going to unconditionally reload it before actually using it.

**Review by aproskuryakov — APPROVED** [review]

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: I'm going to delete this line and change #13993 to address this problem.

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py (hunk 2 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: Will clarify this comment, yes, it's because another process could have (and, in fact, is likely to have) installed the package we're about to install.

File: Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py (hunk 3 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: The thinking here was that we want to safely load this file, but your point about unconditionally reloading is a good one, especially because in the case where we already have all the packages we need installed (which is the most common case), having a lock here would slow down execution. Removing this lock.

Comment (aproskuryakov) [reply Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: My thought was more about not loading the manifest here at all. Do we use it before reloading?

Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitcorepy/webkitcorepy/autoinstall.py]: The short answer is "yes".

The manifest contains a list of all installed libraries. We consult that list to see if we need to install a library and then we need to update the manifest after installing the library. I don't think we need a lock here, and actually, a lock is probably counter-productive.

If someone is in the process of modifying the manifest and the manifest is invalid, that will cause us to treat the manifest as empty. But in that event, we will end up attempting to re-read the manifest (while acquiring the lock) when it comes time to install our first library.

If the manifest is out of date, the same logic applies. All of the libraries we think we have access to are definitely present (since we don't delete libraries), and if we encounter one which a more updated manifest lists, we will find that library when re-reading the manifest.

The only case where we find ourselves in trouble is if two processes define different versions of the same library are running in parallel. That's not a case we can resolve (those two processes should be using different autoinstall locations).
